src/common/VirtualTerminal_FreeBSD.cpp

```diff
@@ -22,16 +23,62 @@
 
 #include "VirtualTerminal.h"
 
+#include <fcntl.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/ioctl.h>
+#include <sys/consio.h>
+
 
 namespace SDDM {
     namespace VirtualTerminal {
+        static char ttyvX[] = "/dev/ttyvX";
+        static const char ttyvs[] = "0123456789ab";
+
         int setUpNewVt() {
-            qDebug() << "New VT is unsupported on FreeBSD";
-            return -1;
+            // there is no way to create VTs on FreeBSD, so
+            // just try to fetch any available
+            return fetchAvailableVt();
         }
 
         void jumpToVt(int vt, bool vt_auto) {
-            qDebug() << "Jumping to VT" << vt << "is unsupported on FreeBSD";
+            int fd = -1;
+
+            for (int i = 0; i < sizeof(ttyvs) / sizeof(char); i++) {
+                ttyvX[9] = ttyvs[i];
+                fd = ::open(ttyvX, O_RDONLY);
+                if (fd != -1)
+                    break;
+                if (fd == -1 && errno == EACCES)
+                    continue;
+                qWarning() << "open(" << ttyvX << ") failed: " << strerror(errno);
+            }
+
+            if (fd == -1) {
+                qWarning() << "Jumping to VT" << vt << " failed: can't open any /dev/ttyvX";
+                return;
+            }
+
+            if (::ioctl(fd, VT_ACTIVATE, vt) == -1) {
+                qWarning() << "ioctl(VT_ACTIVATE) failed: " << strerror(errno);
+            }
+        }
+
+        int fetchAvailableVt() {
+            int fd = ::open("/dev/console", O_RDONLY);
```

Comment:
What if /dev/console isn't a VT (i.e. the primary system console is a serial console)? This will fail, but we probably want to be using /dev/ttyv0 so we can call VT_OPENQRY still?

Reply:
I have minimal understanding of all that terminal stuff. Are you saying that I should change it `/dev/ttyv0` and everything will be OK?

Reply:
For some reason in `jumpToVt()` I iterate over `/dev/ttvyXX` in order to find an actionable VT. Should I do that in every method of `VirtualTerminal_FreeBSD`?

Reply:
At least in ConsoleKit using `ttyv0` instead of `/dev/console` broke things: https://github.com/ConsoleKit2/ConsoleKit2/pull/116#pullrequestreview-786669165

Reply:
Well, /dev/console is a very thin wrapper around whatever the default tty is, which is normally ttyv0 on x86 (it's never going to be ttyv1 or higher, but can be ttyu0), so that's rather surprising. More likely is one of the other files in the list between /dev/ttyv0 and /dev/console gets used instead. Very few VT operations care which window you have open; VT_SETMODE is one (needs to be the one you're trying to modify as it doesn't take a VT argument), as is VT_WAITACTIVE if you give it a VT of 0, which means whichever you have open. It's hard to extract much useful information from that FreeBSD PR, there's very little actual tracing of what went wrong beyond "my desktop broke" and "dropping this patch magically fixed it". All I know is this VirtualTerminal implementation is woefully deficient (you don't even VT_WAITACTIVE after calling VT_ACTIVATE, so you have a nice race there that I reliably lost on my non-x86 system) and it took me very little effort to generalise the existing working Linux VirtualTerminal.cpp to support FreeBSD, which then didn't suffer from those races and worked just fine. https://github.com/CTSRD-CHERI/cheribsd-ports/pull/66/commits/9f33d093f9e30b691d7dd73417d422371c30c9df has the patch I came up with for 0.19.0 that likely isn't hard to adapt to the current develop tree and works reliably in my environment.

Reply:
Awesome, thanks for the patch. I made a PR to upstream it.